src/libraries/System.Private.CoreLib/src/System/String.Comparison.cs

```diff
@@ -680,32 +682,117 @@ public bool Equals([NotNullWhen(true)] string? value, StringComparison compariso
             }
         }
 
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static ulong ReadUInt64(string str) =>
+            Unsafe.ReadUnaligned<ulong>(ref Unsafe.As<char, byte>(ref str._firstChar));
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static ulong ReadUInt64(string str, nint offset) =>
+            Unsafe.ReadUnaligned<ulong>(ref Unsafe.As<char, byte>(ref Unsafe.Add(ref str._firstChar, offset)));
+
         // Determines whether two Strings match.
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
         public static bool Equals(string? a, string? b)
         {
-            // Transform 'str == ""' to 'str != null && str.Length == 0' if either a or b are jit-time
-            // constants. Otherwise, these two blocks are eliminated
-            if (RuntimeHelpers.IsKnownConstant(a) && a != null && a.Length == 0)
+#if TARGET_64BIT && !MONO && !BIGENDIAN
+            // Try to unroll Equals in case of constant 'b' for b.Length in [0..16] range
+            if (RuntimeHelpers.IsKnownConstant(b) && !RuntimeHelpers.IsKnownConstant(a) && b != null)
             {
-                return b != null && b.Length == 0;
+                // Unroll using SWAR
+                if (b.Length <= 8)
+                {
+                    return EqualsUnrolled_0_to_8(a, b);
+                }
+                // Unroll using two Vector128s
+                if (b.Length >= 9 && b.Length <= 16 && Vector128.IsHardwareAccelerated)
+                {
+                    return EqualsUnrolled_9_to_16(a, b);
+                }
+                // NOTE: for some values we can emit a more optimal codegen e.g. for Length 7 and 8
+                // we can use a single Vector128 or add Vector256 path for Length in [16..32] range
+                // but we need to be careful here with inliner's budget
             }
 
-            if (RuntimeHelpers.IsKnownConstant(b) && b != null && b.Length == 0)
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            static bool EqualsUnrolled_0_to_8(string? a, string b)
             {
-                return a != null && a.Length == 0;
-            }
+                if (a != null)
+                {
+                    if (b.Length == 0)
+                    {
+                        // Fold Equals(a, "") to 'a != null && a.Length == 0'
+                        return a.Length == 0;
+                    }
+                    if (b.Length == 1)
+                    {
+                        // Load Length, ch1, \0 into ulong (so we can skip the Length check)
+                        return ReadUInt64(a, -2) == (((ulong)b[0] << 32) | 1UL);
```

Comment:
Is this `Read(a)` safe without a length check?

Reply:
I don't know if I can safely rely on the fact that even an empty string actually has 4 byte Length, 2 byte \0 and 2 byte padding 🤔  (code is guarded for 64bit only)

Reply:
Right, forgot about the padding